### PR TITLE
feat: Add support for vectorized UDFs

### PR DESCRIPTION
**Pull request by rtpsw**
For https://github.com/ibis-project/ibis-substrait/issues/236

### CODE REVIEW

**Comment by rtpsw** [comment]
@cpcloud - I see the `AttributeError: module 'ibis.expr.operations' has no attribute 'VectorizedUDF'` is recurring. Could you advise what it means or how to approach fixing it? It seems to be related to the nix setup, which I don't know much about.

**Comment by cpcloud** [comment]
That class isn't a public API so you'll have to access it from the module where it's defined, not `ibis.expr.operations`. I don't think this is related to the nix setup.

**Comment by rtpsw** [comment]
Ah, I see now. This should get fixed once https://github.com/ibis-project/ibis/pull/4029 is finished, so let's focus on that first.

File: ibis_substrait/compiler/core.py [hunk 1]
```diff
@@ -64,18 +65,39 @@ def function_id(self, expr: ir.ValueExpr) -> int:
...
+            ] = ste_decl.ExtensionFunction(
                 extension_uri_reference=self.extension_uri.extension_uri_anchor,
                 function_anchor=next(self.id_generator),
                 name=op_name,
+                udf=None
+                if not isinstance(op, VectorizedUDF)
+                else (
+                    ste_decl.ExtensionFunction.UserDefinedFunction(
+                        code=base64.b64encode(
+                            cloudpickle.dumps(op.func.__wrapped__)
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply ibis_substrait/compiler/core.py:93]: I locally checked by adding a no-op decorator and the pickling works.

File: ibis_substrait/compiler/core.py [hunk 2]
```diff
@@ -64,18 +65,39 @@ def function_id(self, expr: ir.ValueExpr) -> int:
             This is a unique identifier for a given operation type, *argument
             types N-tuple.
         """
+        from .translate import translate
+
         op = expr.op()
-        op_type = type(op)
-        op_name = IBIS_SUBSTRAIT_OP_MAPPING[op_type.__name__]
+        if isinstance(op, VectorizedUDF):
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply ibis_substrait/compiler/core.py:71]: I'll clean it up.

File: proto/substrait/extensions/extensions.proto [hunk 3]
```diff
@@ -63,8 +65,29 @@ message SimpleExtensionDeclaration {
...
+        string summary = 2;
+
+        // Description of the user-defined function.
+        string description = 3;
+
+        // Input types of the user-defined function.
+        repeated Type input_types = 4;
+
+        // Output type of the user-defined function.
+        Type output_type = 5;
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply proto/substrait/extensions/extensions.proto:89]: Yes. I'll create a PR upstream once I locally deal with the comments and check the code.

File: ibis_substrait/compiler/core.py [hunk 4]
```diff
@@ -64,18 +65,39 @@ def function_id(self, expr: ir.ValueExpr) -> int:
...
+            op_name = op.func.__wrapped__.__name__
+            import base64
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply ibis_substrait/compiler/core.py]: Done.

File: ibis_substrait/compiler/core.py [hunk 5]
```diff
@@ -64,18 +65,39 @@ def function_id(self, expr: ir.ValueExpr) -> int:
...
+
+            import cloudpickle
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply ibis_substrait/compiler/core.py]: Done.

File: ibis_substrait/compiler/core.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rtpsw) [reply ibis_substrait/compiler/core.py]: Done.

**Comment by rtpsw** [comment]
@cpcloud, I used `poetry add cloudpickle` and it updated a bunch of other dependencies, which I'm not sure is desirable.

**Comment by rtpsw** [comment]
> Note that I will move the protobuf changes in this PR upstream.

https://github.com/substrait-io/substrait/pull/234

**Comment by icexelloss** [comment]
We can cancel this one. Vectorized UDF is supported (elementwise)

**Pull request closed without merging**